narrow: Add starred messages narrow

PIG208 opened the pull request:
Unreads can probably be handled in a separate PR, because it will probably grow in complexity.

This is stacked on top of #869

PIG208 commented:
This does not implement live updates for newly starred/unstarred messages. We should able to handle #818 together with this.

PIG208 reviewed (commented):
We will leave unreads for starred messages for now, as it is pretty uncommon to have unreads for them.

PIG208 commented:
Skipping the first 3 commits, this should be reviewable independently from #869.

chrisbobbe commented:
Looks like CI is failing.

PIG208 commented:
Should be fixed now. It was an error emerged from moved messages support across narrows.

chrisbobbe commented:
Thanks! Ah, but it looks like some conflicts appeared after you pushed that revision; could you resolve those please? :)

PIG208 commented:
Thanks! Just rebased it.

chrisbobbe commented on lib/model/internal_link.dart:
```diff
@@ -203,6 +203,7 @@ Narrow? _interpretNarrowSegments(List<String> segments, PerAccountStore store) {
       case IsOperand.private:
       case IsOperand.alerted:
       case IsOperand.starred:
+        return const StarredMessagesNarrow();
```
```
internal_link: Parse "/is/starred" internal links.
```

This is a bug. :) Since cases before `IsOperand.starred` are empty, they'll also give `const StarredMessagesNarrow()`, just like the `IsOperand.starred` case.

https://dart.dev/language/branches#switch-statements

> Non-empty `case` clauses jump to the end of the switch after completion. They do not require a `break` statement.

It's a common pattern to want multiple cases to give the same thing, and to express that with empty `case` clauses, as here.

To fix (here and in the test, which has the same bug), I would just reorder:

```dart
    switch (isElementOperands.single) {
      case IsOperand.mentioned:
        return const MentionsNarrow();
      case IsOperand.starred:
        return const StarredMessagesNarrow();
      case IsOperand.dm:
      case IsOperand.private:
      case IsOperand.alerted:
      case IsOperand.followed:
      case IsOperand.resolved:
      case IsOperand.unread:
      case IsOperand.unknown:
        return null;
    }
```

chrisbobbe reviewed (changes requested):
Thanks! Small comments below, including one about a bug I noticed.

In manual testing, I noticed that when I'm viewing the starred-messages narrow, when I unstar a message, it remains in the list. At first I thought this was a bug, but Greg pointed out that it's actually a product decision; it lets you easily undo the unstar action if you want. So that seems fine.

—But let's put a comment in the code, clarifying that it's intentional (to offer undo), so readers don't have to wonder. This point already applies for messages with @-mentions in the mentions narrow, I think, so it would be good to mention that narrow too, in a separate, earlier commit. I think the right place to comment on these things would be in `MessageStoreImpl.handleUpdateMessageFlagsEvent`, near where it operates on the message-list views (by having them notify listeners but not remove messages).

…Hmm. Thinking more, actually…The current behavior for remove-star events does seem helpful, but wouldn't it also be helpful if `MessageStoreImpl.handleUpdateMessageFlagsEvent` acted on _add_-star events by making the message appear in the starred-narrow message list? Similarly for mentioned messages in the mentions narrow.

PIG208 commented:
Regarding

> …Hmm. Thinking more, actually…The current behavior for remove-star events does seem helpful, but wouldn't it also be helpful if MessageStoreImpl.handleUpdateMessageFlagsEvent acted on add-star events by making the message appear in the starred-narrow message list? Similarly for mentioned messages in the mentions narrow.

Live-updates for starred messages and @-mentions are post-launch features, so we leave them out for now:

> This does not implement live updates for newly starred/unstarred messages. We should able to handle #818 together with this as a post-launch follow-up.

PIG208 commented:
Having comments for now should be suitable.

PIG208 commented:
Updated the PR. Thanks for the review!

chrisbobbe commented:
Thanks, LGTM! Over to Greg, then.

gnprice reviewed (commented):
Thanks @PIG208, and thanks @chrisbobbe for the previous review!

Generally this looks great. Comments below on the tests, mostly just on their structure.

Separately, I think the last three commits make most sense squashed together:
8d8841ee9 narrow: Add starred messages narrow.
3b3b4c389 internal_link: Parse "/is/starred" internal links.
bf0759dea nav: Add a button for starred messages page.

The latter two are small changes adding pieces that really should be there in order to say we support the starred-messages narrow. It seems like the main difference between those and most of the changes in the main commit is just that the main commit has the changes that are prompted by exhaustive-switch checking once the StarredMessagesNarrow type is added. That's a very useful tool but doesn't need to decide our commit structure.

(Even in the other direction: if the changes required to satisfy exhaustive `switch`es added up to a big change that we didn't want to put in a single commit, we could split it up by making some of the new cases just `throw UnimplementedError()` or the like, with TODO comments pointing to the relevant issue.)

PIG208 commented:
Updated the PR. Thanks for the review!

Made separate functions for one test each in action_sheet_test.dart and message_list_test.dart, and removed the extra regression test. I also squashed the later commits together as suggested.

PIG208 commented on test/widgets/message_list_test.dart:
```diff
@@ -123,31 +123,25 @@ void main() {
-
-      final element = tester.element(find.byType(CodeBlock));
-      final padding = MediaQuery.of(element).padding;
-      check(padding).equals(EdgeInsets.zero);
```
Yeah, this interpretation makes sense me. I decided to remove the `MentionsNarrow` test, because I think they will be largely verifying the same code path. It is reasonable to check for the absence of compose box in different narrows, but we are more interested in testing the inset fix here.

PIG208 commented on test/widgets/message_list_test.dart:
```diff
@@ -717,40 +711,34 @@ void main() {
-        await tester.pump();
-        check(findInMessageList('stream name')).length.equals(1);
-        check(findInMessageList('topic name')).length.equals(1);
-      });
```
Thanks. `sticky_header_test.dart` is an interesting example. Dropped the parameterization commit.

gnprice commented on test/widgets/action_sheet_test.dart:
```diff
@@ -388,10 +388,16 @@ void main() {
 
     testWidgets('not offered in MentionsNarrow (composing to reply is not yet supported)', (tester) async {
-      final message = eg.streamMessage();
+      final message = eg.streamMessage(flags: [MessageFlag.mentioned]);
       await setupToMessageActionSheet(tester, message: message, narrow: const MentionsNarrow());
```
action_sheet test: Create message with the appropriate flag.
    
    The message wouldn't otherwise appear in MentionsNarrow.

Hmm indeed. But I'm a bit puzzled then: why wasn't this test already failing?

Ah it's because the code actually *will* show the message. So this is just a realism fix, making the simulated server response something a non-buggy server might give on this request.

gnprice commented on lib/model/message.dart:
```diff
@@ -270,6 +270,14 @@ class MessageStoreImpl with MessageStore {
       if (anyMessageFound) {
         for (final view in _messageListViews) {
           view.notifyListenersIfAnyMessagePresent(event.messages);
+          // TODO(#818): Support MentionsNarrow live-updates when handling
+          // @-mentioned flags.
```
nit: indent any continuation lines on a TODO comment:
```suggestion
          // TODO(#818): Support MentionsNarrow live-updates when handling
          //   @-mentioned flags.
```

gnprice replied:
(also "@-mentioned flags" reads to me as about flags that were @-mentioned; could say "mention flags" or "@-mention flags")

gnprice reviewed (commented):
Thanks for the revision!

A few new nits below — otherwise all looks good. I'll fix those up and merge.

gnprice commented on test/widgets/message_list_test.dart:
```diff
@@ -135,19 +135,6 @@ void main() {
       check(padding).equals(EdgeInsets.zero);
     });
-
-    testWidgets('content in MentionsNarrow not asked to consume insets (including bottom)', (tester) async {
```
Oh one other commit-message comment:

    A test on CombinedFeedNarrow is sufficient for preventing this particular
    regression. We could go further and test this on other narrows, but
    there is little to gain other than verifying that those other narrows do
    not have a compose box, which is covered statically with the exhaustive
    check from hasComposeBox.

The last sentence here doesn't sound right, as this test didn't check that the narrow doesn't have a compose box. Rather, if the narrow had a compose box, the test would pass even if the bug were to return on other narrows that don't have the compose box. The original bug reproduced only on CombinedFeedNarrow (the only narrow lacking a compose box at the time), which is why the repro test used that narrow.

gnprice commented:
> I'll fix those up and merge.

Done. I also added a commit to better pin down that existing regression test's assumptions:
5bcb5d4ce msglist test [nfc]: Pin down lack of compose box in no-insets regression test
